### PR TITLE
Implement DOMStringMap::SupportedPropertyNames and NamedNodeMap::SupportedPropertyNames #7695

### PR DESCRIPTION
**Pull request by nfallen**
Here is a draft for issue #7695 with web platform tests. 
Thanks for reviewing!
https://dom.spec.whatwg.org/#namednodemap
https://html.spec.whatwg.org/multipage/infrastructure.html#domstringmap

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8114)

### CODE REVIEW

**Comment by jdm** [comment]
Our automated CI server reports:
```
./components/script/dom/htmlelement.rs:282: trailing whitespace
./components/script/dom/htmlelement.rs:283: links to WHATWG single-page url, change to multi page: https://html.spec.whatwg.org/multipage/#attr-data-
./components/script/dom/namednodemap.rs:93: trailing whitespace

The command "./mach test-tidy" exited with 1.
```

File: components/script/dom/htmlelement.rs [hunk 1]
```diff
@@ -278,6 +278,39 @@ fn to_snake_case(name: DOMString) -> DOMString {
     attr_name
 }
 
+/*
+ * https://html.spec.whatwg.org/multipage/#attr-data-
+ * if this attribute is in snake case with a data- prefix,
+ * this function returns a name converted to camel case
+ */
+fn to_camel_case(name: DOMString) -> Option<DOMString> {
+    let has_uppercase = name.chars().fold(false, |exists_rest, curr_char| {
```
Comment (dzbarsky) [inline components/script/dom/htmlelement.rs]: Would be a little nicer if you used any or all (http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all)

File: components/script/dom/htmlelement.rs [hunk 2]
```diff
@@ -278,6 +278,39 @@ fn to_snake_case(name: DOMString) -> DOMString {
...
+        curr_char.is_uppercase() | exists_rest
+    });
+    if !name.starts_with("data-") || has_uppercase {
+        None
+    } else {
+        let mut result = "".to_owned();
+        let mut name_chars = name.chars();
+        //iterate through prefix "data-"
+        for _ in 0..5 {
+            let _ = name_chars.next();
```
Comment (dzbarsky) [inline components/script/dom/htmlelement.rs]: http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip

File: components/script/dom/htmlelement.rs [hunk 3]
```diff
@@ -302,6 +335,20 @@ impl HTMLElement {
         let local_name = Atom::from_slice(&to_snake_case(local_name));
         element.remove_attribute(&ns!(""), &local_name);
     }
+
+    pub fn supported_prop_names_custom_attr(&self) -> Vec<DOMString> {
+        let element = ElementCast::from_ref(self);
+        element.attrs().iter().map(JS::root).map(|attr| {
+            let raw_name = (**attr.r().local_name()).to_owned();
+            to_camel_case(raw_name)
+        }).fold(Vec::new(), |mut acc, opt| {
```
Comment (dzbarsky) [inline components/script/dom/htmlelement.rs]: I think you may be able to use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map and collect the results

**Comment by dzbarsky** [comment]
Thanks for the PR!

**Comment by nfallen** [comment]
Thanks for the comments @dzbarsky! Updating my pull request to use the functions you suggested.

File: components/script/dom/htmlelement.rs [hunk 4]
```diff
@@ -278,6 +278,36 @@ fn to_snake_case(name: DOMString) -> DOMString {
...
+    );
+    if !name.starts_with("data-") || has_uppercase {
+        None
+    } else {
+        let mut result = "".to_owned();
+        //Skip the prefix
+        let mut name_chars = name.chars().skip(5);
+        while let Some(curr_char) = name_chars.next() {
+            if curr_char == '\x2d' {
+                let next_char_opt = name_chars.next();
```
Comment (dzbarsky) [inline components/script/dom/htmlelement.rs]: This may be nicer as an if let; you'll be able to drop the match and the break.

**Comment by nox** [comment]
Thanks for your contribution!

-S-awaiting-review +S-needs-code-changes 


---

Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/domstringmap.rs, line 67 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1B9AuhfJsBAdX2XXyE)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/domstringmap.rs#L67)):</sup>
No need to root `self.element` anymore.


---

<sup>**[components/script/dom/domstringmap.rs, line 70 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BChY-UrSY_c6xokk5)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/domstringmap.rs#L70)):</sup>
I'm pretty sure you don't need to repeat the type here. No need to call `DOMString::clone` yourself either:

```rust
self.element.supported_prop_names_custom_attr().iter().cloned().collect()
```


---

<sup>**[components/script/dom/htmlelement.rs, line 292 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BE2TRnH-JVJ7qt433)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L292)):</sup>
Nit: Either use `//` or do a proper `///` documentation block.


---

<sup>**[components/script/dom/htmlelement.rs, line 293 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGsFzbs_fpzgoz8e1)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L293)):</sup>
This should take `&str`.


---

<sup>**[components/script/dom/htmlelement.rs, line 295 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGA9WZ8k2BDzxbqaY)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L295)):</sup>
This should compare `curr_char` to `curr_char.to_ascii_uppercase()`, only ASCII uppercase characters are forbidden.


---

<sup>**[components/script/dom/htmlelement.rs, line 310 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BFtVuDw8ymsMAeG5Q)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L310)):</sup>
Only hyphens followed by ASCII lowercase characters should be transformed, if the hyphen is followed by anything else, it shouldn't.


---

<sup>**[components/script/dom/htmlelement.rs, line 347 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGfbrsqO-2D7j9QpS)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L347)):</sup>
I bitrot this, now it reads as `self.upcast::<Element>()`.


---

<sup>**[components/script/dom/htmlelement.rs, line 349 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGoCh5Mkk50-U1fcu)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L349)):</sup>
Why do you need to clone the name first? 


---

<sup>**[components/script/dom/htmlelement.rs, line 287 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8114#-K17Pm291oQmT7z6kh4F-r3-287)** ([raw file](https://github.com/servo/servo/blob/9c44cd853fccf0ea029d83c4654217f76a9fc602/components/script/dom/htmlelement.rs#L287)):</sup>
Yes, please use that.


---

<sup>**[components/script/dom/htmlelement.rs, line 297 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8114#-K17Pm291oQmT7z6kh4F-r3-297)** ([raw file](https://github.com/servo/servo/blob/9c44cd853fccf0ea029d83c4654217f76a9fc602/components/script/dom/htmlelement.rs#L297)):</sup>
Actually, you should just do `let mut name_chars = name[6..].chars()`.


---

<sup>**[components/script/dom/htmlelement.rs, line 344 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8114#-K17Pm291oQmT7z6kh4F-r3-344)** ([raw file](https://github.com/servo/servo/blob/9c44cd853fccf0ea029d83c4654217f76a9fc602/components/script/dom/htmlelement.rs#L344)):</sup>
Yes, please use `filter_map`.


---

<sup>**[components/script/dom/htmlelement.rs, line 298 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8114#-K17Pm291oQmT7z6kh4F-r4-298)** ([raw file](https://github.com/servo/servo/blob/66008c226dde789c8b3a54fdeb103b540ed72f43/components/script/dom/htmlelement.rs#L298)):</sup>
You shouldn't be breaking if there is no next character anyway, because in this case the hyphen should be kept as is.

```rust
while let Some(curr_char) = name_chars.next() {
    if curr_char == '\x2d' {
        if let Some(next_char) = name_chars.next();
            result.extend(next_char.…);
            continue;
        }
    }
    result.push(curr_char);
}
```


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nox** [comment]
Sorry for some remarks that were outdated with regard to the latest code you pushed, but please do address the others. :)

Note that there is a merge commit in your branch that has nothing to do here.


---

Reviewed 6 of 6 files at r1, 4 of 4 files at r2, 4 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BHEJHpI17cLGumBtJ)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L289)):</sup>
Use the multipage link please (add `multipage/` at the beginning of the path).


---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BHf5Wj9yN1CuMIWKX)** ([raw file](https://github.com/servo/servo/blob/66008c226dde789c8b3a54fdeb103b540ed72f43/components/script/dom/htmlelement.rs#L289)):</sup>
Please use curly braces when doing a multiline lambda.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nfallen** [comment]
I updated my code based on the comments. As of now if an attribute is just called "data-" with no characters following the code is returning nothing, but I could change it to be an empty string. I am unclear on what the behavior should be based on the specs: https://html.spec.whatwg.org/multipage/#attr-data-*

**Comment by nox** [comment]
@nfallen The spec says:

> A custom data attribute is an attribute in no namespace whose name starts with the string "data-", has at least one character after the hyphen, is XML-compatible, and contains no uppercase ASCII letters.

So "data-" shouldn't be included.

And yes, you should rebase, not merge.

**Comment by nox** [comment]
And if you just have some fixes to do, without conflicts, you need neither, just amend the commits on your branch accordingly.

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes +S-needs-rebase 


---

Reviewed 1 of 1 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 297 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8114#-K17Pm291oQmT7z6kh4F-r3-297)** ([raw file](https://github.com/servo/servo/blob/9c44cd853fccf0ea029d83c4654217f76a9fc602/components/script/dom/htmlelement.rs#L297)):</sup>
Oh right, 5… Off-by-one errors are my nightmare. :P


---

<sup>**[components/script/dom/htmlelement.rs, line 283 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1Izct4A722-qJmVqbi)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L283)):</sup>
Nit: period at the end, and one space instead of two after `//`. 


---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-tJn0EJBKMQ6lEqG)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L289)):</sup>
I'm pretty sure you don't need to check the length, given you check whether the name starts with `data-`. 


---

<sup>**[components/script/dom/htmlelement.rs, line 290 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J00idwFxDQOlGPgkP)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L290)):</sup>
Remove that.


---

<sup>**[components/script/dom/htmlelement.rs, line 342 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-3yPKxo5MUTpICjY)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L342)):</sup>
Remove the parens.


---

<sup>**[components/script/dom/htmlelement.rs, line 344 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-6AqzPxYUna_OQyN)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L344)):</sup>
Remove the type.


---

<sup>**[components/script/dom/namednodemap.rs, line 92 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-7bsuR5NGAB6lYqH)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/namednodemap.rs#L92)):</sup>
Remove the type.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nfallen** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 295 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGA9WZ8k2BDzxbqaY)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L295)):</sup>
Since curr_char and curr_char.to_ascii_uppercase() are also equal if the character is a hyphen or something else that is not a-z, I can use curr_char.is_ascii() && curr_char.is_uppercase().


---

<sup>**[components/script/dom/htmlelement.rs, line 310 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BFtVuDw8ymsMAeG5Q)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L310)):</sup>
Thanks for pointing this out. As a follow up clarification on the specs, the definition of a custom data attribute says there must be at least one character after the "data-" prefix, but the section about getting the list of name-value pairs says to add "each content attribute on the element whose first five characters are the string "data-" and whose remaining characters (if any) do not include any uppercase ASCII letters". Should I be returning an empty string if there is an attribute that is simply "data-", or leaving that string out?


---

<sup>**[components/script/dom/htmlelement.rs, line 349 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGoCh5Mkk50-U1fcu)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L349)):</sup>
You're right, I don't need to. I can just borrow it!


---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-tJn0EJBKMQ6lEqG)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L289)):</sup>
I am checking the length to make sure there is at least one character after the "data-" prefix, so that I can return None instead of the empty string if not.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nox** [comment]
-S-awaiting-review -S-needs-rebase +S-needs-code-changes +I-spec-unclear 


---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 295 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGA9WZ8k2BDzxbqaY)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L295)):</sup>
Good call.


---

<sup>**[components/script/dom/htmlelement.rs, line 310 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BFtVuDw8ymsMAeG5Q)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L310)):</sup>
I'll look into that.


---

<sup>**[components/script/dom/htmlelement.rs, line 349 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1BGoCh5Mkk50-U1fcu)** ([raw file](https://github.com/servo/servo/blob/d0d6c4d3b1025e532d3ac0b294559741b95b531f/components/script/dom/htmlelement.rs#L349)):</sup>
That's still a clone. Just pass it as `&str` to `to_camel_case`.


---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-tJn0EJBKMQ6lEqG)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L289)):</sup>
You can still do this like that:

```rust
if !name.starts_with("data-") {
    return None;
}
let name = &name[5..];
if name.is_empty() {
    return None;
}
let has_uppercase = …;
if has_uppercase {
    return None;
}
let mut result = "".to_owned();
let mut name_chars = name.chars();
…
```

No?

Don't wrap the non-error path in an `else` branch, that's generally considered bad style.


---

<sup>**[components/script/dom/htmlelement.rs, line 344 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-6AqzPxYUna_OQyN)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L344)):</sup>
Did it fail when removing it completely? I.e. just `.collect()`.


---

<sup>**[components/script/dom/namednodemap.rs, line 92 \[r7\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1KykqywuHZU_0x5-rB)** ([raw file](https://github.com/servo/servo/blob/31302548071239e7218e63fb8c08c1f8fb8f9e91/components/script/dom/namednodemap.rs#L92)):</sup>
Just `.collect()` didn't work?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nox** [comment]
-I-spec-unclear 


---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1J-tJn0EJBKMQ6lEqG)** ([raw file](https://github.com/servo/servo/blob/caa28d200c5b17b524763546c1cb955dd5bf0dff/components/script/dom/htmlelement.rs#L289)):</sup>
After further inquiry, you can drop the length check, the thing you quoted is an author requirement, as opposed to an UA requirement. Still, don't forget to rewrite the code as suggested, so that we don't even check whether "data-" is uppercase or not.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nfallen** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/namednodemap.rs, line 92 \[r7\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1KykqywuHZU_0x5-rB)** ([raw file](https://github.com/servo/servo/blob/31302548071239e7218e63fb8c08c1f8fb8f9e91/components/script/dom/namednodemap.rs#L92)):</sup>
No, you're right, it works! Thanks for clarifying


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nox** [comment]
Almost here!

-S-awaiting-review +S-needs-code-changes 


---

Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 344 \[r8\]](https://reviewable.io:443/reviews/servo/servo/8114#-K1PE6GZCHIsE1w1V3aB)** ([raw file](https://github.com/servo/servo/blob/498750968c4b9b24cd71f62b11f78e935bdcf56d/components/script/dom/htmlelement.rs#L344)):</sup>
Mmmh, the method `Attr::local_name` returns an `Atom`, does this actually compile? Shouldn't it be `to_camel_case(&raw_name)`?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by nox** [comment]
Please squash everything together when you fix that last remark. :)

**Comment by nfallen** [comment]
Uh oh I definitely messed up somewhere when trying to squash

**Comment by nox** [comment]
@nfallen Yes. Feel free to ping me on IRC if you need help. This should do the trick:

```
git checkout -b tmp master
git cherry-pick 17fc4e0
```

**Comment by nfallen** [comment]
Just wanted to check in on this. I squashed and rebased with master, so is there anything else I should do to make this PR ready to merge?

**Comment by eefriedman** [comment]
GitHub doesn't send a notification when you modify a branch, so it's usually a good idea to write a short note on the PR when you want a reviewer to look again.


---

Reviewed 1 of 4 files at r3, 1 of 4 files at r6, 2 of 3 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8114)

**Comment by mbrubeck** [comment]
You can simply delete this file now that these tests are passing:

[tests/wpt/metadata/html/dom/elements/global-attributes/dataset-enumeration.html.ini](https://github.com/servo/servo/blob/master/tests/wpt/metadata/html/dom/elements/global-attributes/dataset-enumeration.html.ini)

```
Unexpected Results
==================

/html/dom/elements/global-attributes/dataset-enumeration.html
-------------------------------------------------------------
PASS expected FAIL A dataset should be enumeratable.
PASS expected FAIL Only attributes who qualify as dataset properties should be enumeratable in the dataset.
```

**Comment by nfallen** [comment]
Deleted the file and updated my commit, thanks @mbrubeck!

**Pull request merged**